src/EditorFeatures/Core/Tagging/AsynchronousTaggerProvider.cs

```diff
@@ -3,79 +3,189 @@
         }
 
-        ITagger<T> ITaggerProvider.CreateTagger<T>(ITextBuffer buffer)
+        private IAccurateTagger<T> GetOrCreateTagger<T>(ITextView textViewOpt, ITextBuffer subjectBuffer) where T : ITag
+        {
+            if (!subjectBuffer.GetOption(EditorComponentOnOffOptions.Tagger))
+            {
+                return null;
+            }
+
+            var tagSource = GetOrCreateTagSource(textViewOpt, subjectBuffer);
+            return tagSource == null
+                ? null
+                : new Tagger(this.asyncListener, this.notificationService, tagSource, subjectBuffer) as IAccurateTagger<T>;
+        }
+
+        private TagSource GetOrCreateTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            TagSource tagSource;
+            if (!this.TryRetrieveTagSource(textViewOpt, subjectBuffer, out tagSource))
+            {
+                tagSource = this.CreateTagSource(textViewOpt, subjectBuffer);
+                if (tagSource == null)
+                {
+                    return null;
+                }
+
+                this.StoreTagSource(textViewOpt, subjectBuffer, tagSource);
+                tagSource.Disposed += (s, e) => this.RemoveTagSource(textViewOpt, subjectBuffer);
+            }
+
+            return tagSource;
+        }
+
+        private bool TryRetrieveTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer, out TagSource tagSource)
+        {
+            return textViewOpt != null
+                ? textViewOpt.TryGetPerSubjectBufferProperty(subjectBuffer, uniqueKey, out tagSource)
+                : subjectBuffer.Properties.TryGetProperty(uniqueKey, out tagSource);
+        }
+
+        private void RemoveTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            if (textViewOpt != null)
+            {
+                textViewOpt.RemovePerSubjectBufferProperty<TagSource, ITextView>(subjectBuffer, uniqueKey);
+            }
+            else
+            {
+                subjectBuffer.Properties.RemoveProperty(uniqueKey);
+            }
+        }
+
+        private void StoreTagSource(ITextView textViewOpt, ITextBuffer subjectBuffer, TagSource tagSource)
         {
-            return _taggerImplementation.CreateTagger<T>(buffer);
+            if (textViewOpt != null)
+            {
+                textViewOpt.AddPerSubjectBufferProperty(subjectBuffer, uniqueKey, tagSource);
+            }
+            else
+            {
+                subjectBuffer.Properties.AddProperty(uniqueKey, tagSource);
+            }
+        }
+
+        public virtual SnapshotPoint? GetCaretPoint(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            // Use 'null' to indicate that the tagger should get the default caret position.
+            return null;
+        }
+
+        public virtual IEnumerable<SnapshotSpan> GetSpansToTag(ITextView textViewOpt, ITextBuffer subjectBuffer)
```

Comment:
Why isn't this a default implementation that does the Right Thing? Anybody who wants to do something like call base in their implementation and tweak the result might be surprised by this.